### PR TITLE
Added the component specification

### PR DESCRIPTION
**Pull request by sarahmaddox**
fixes https://github.com/kubeflow/website/issues/570

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/website/604)

### CODE REVIEW

**Comment by sarahmaddox** [comment]
Preview: https://deploy-preview-604--competent-brattain-de2d6d.netlify.com/docs/pipelines/reference/component-spec/

**Comment by sarahmaddox** [comment]
/assign @Ark-kun 
/assign @abhi-g 

Note: I'll link this page to the component development guide https://github.com/kubeflow/website/pull/601 (and vice versa) once both PRs are merged.

File: content/docs/pipelines/reference/component-spec.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+* In `component.yaml`: `{inputValue: Rounds}`
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
+
+### Outputs
+
+Output paths are filled in by the pipeline system. The Output construct is
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
+* In `component.yaml': `{outputPath: Trained model}`
+* Resulting Argo YAML: `"/outputs/trained_model/data"`
```
Comment (paveldournov) [inline content/docs/pipelines/reference/component-spec.md]: is trained_model directory constructed from the "Trained model" name? In the docs - can we always keep the names matching? Or we need to reference a clear translation logic here.

File: content/docs/pipelines/reference/component-spec.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    * `name`: Human-readable name of the input/output. Name must be
+        unique inside the inputs or outputs section, but an output may have the
+        same name as an input.
+    * `description`: Human-readable description of the input/output.
+    * `default`: Specifies the default value for an input. Only
+        valid for inputs.
+    * `type`: Specifies the type of input/output. The types are used
+        as hints for pipeline authors and can be used by the pipeline system/UI
+        to validate arguments and connections between components. Basic types
+        are **String**, **Integer**, **Number**, **Boolean**.
```
Comment (paveldournov) [inline content/docs/pipelines/reference/component-spec.md]: Is this a full list of supported types? If not - can we have a reference here to the full list?

Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 1 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: @Ark-kun

Comment (Ark-kun) [reply content/docs/pipelines/reference/component-spec.md]: >is trained_model directory constructed from the "Trained model" name?

Yes.

> can we always keep the names matching?

We'll fix this part to make the names match. This is only a path though and the user does not interact with it.

> Or we need to reference a clear translation logic here.

We can describe the translation logic. But half of the purpose of generating the paths (as opposed to hard-coding the paths) is that it's opaque and can be changed later. We could as well just generate UUID paths like `/outputs/5c9b2d651e6c01cb41d834faa77925157d5d1599/data`. The user should not really care about or depend on some concrete logic.

Comment (Ark-kun) [reply content/docs/pipelines/reference/component-spec.md]: Maybe we should not mention Argo here at all.

File: content/docs/pipelines/reference/component-spec.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+You can set all other Kubernetes container properties when you
+use the component inside a pipeline.
+
+## Using placeholders for command-line arguments
+
+### Consuming input by value (parameter)
+
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
```
Comment (Ark-kun) [inline content/docs/pipelines/reference/component-spec.md]: In `component.yaml`:
```command: [program.py, --rounds, {inputValue: Rounds}]```
In pipeline code:
```task1 = component1(rounds=150)```
Resulting command-line (the placeholder is substituted):
```program.py --rounds 150```

File: content/docs/pipelines/reference/component-spec.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
+
+### Outputs
+
+Output paths are filled in by the pipeline system. The Output construct is
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
+* In `component.yaml`: `{outputPath: Trained model}`
```
Comment (Ark-kun) [inline content/docs/pipelines/reference/component-spec.md]: In `component.yaml`:
```command: [program.py, --out-model, {outputPath: trained_model}]```
In pipeline code:
```task1 = component1()```
Resulting command-line (the placeholder is substituted with generated path):
```program.py --out-model /outputs/trained_model/data```

File: content/docs/pipelines/reference/component-spec.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+### Consuming input by value (parameter)
+
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
+
+### Outputs
+
+Output paths are filled in by the pipeline system. The Output construct is
```
Comment (Ark-kun) [inline content/docs/pipelines/reference/component-spec.md]:
```suggestion
Output paths are filled in by the pipeline system. The outputPath placeholder is
```

File: content/docs/pipelines/reference/component-spec.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
```
Comment (Ark-kun) [inline content/docs/pipelines/reference/component-spec.md]:
```suggestion
The parent directories of the path might or might not not exist.
The program must handle both cases without errors.

```

File: content/docs/pipelines/reference/component-spec.md (hunk 1 above)
Comment (Ark-kun) [reply content/docs/pipelines/reference/component-spec.md]: I've added comments with replacement text. We should not mention Argo. (Remove this line)

File: content/docs/pipelines/reference/component-spec.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+use the component inside a pipeline.
+
+## Using placeholders for command-line arguments
+
+### Consuming input by value (parameter)
+
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
```
Comment (Ark-kun) [inline content/docs/pipelines/reference/component-spec.md]:
```suggestion
```

File: content/docs/pipelines/reference/component-spec.md (hunk 3 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Thanks. Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 4 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 5 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 6 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 7 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

File: content/docs/pipelines/reference/component-spec.md (hunk 1 above)
Comment (sarahmaddox) [reply content/docs/pipelines/reference/component-spec.md]: Done.

**Comment by sarahmaddox** [comment]
Thanks for the review. :) I've addressed all the comments. PTAL https://deploy-preview-604--competent-brattain-de2d6d.netlify.com/docs/pipelines/reference/component-spec/

**Pull request merged**